File: d14/puzzle2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <unordered_map>
#include <utility>
#include <cmath>
#include <deque>
using namespace std;

#define height 103
#define width 101
void render(vector<vector<deque<int>>>& map){
    for (int i = 0; i < height; i++){
        for (int j = 0; j < width; j++){
            if (map[i][j].empty())
                cout << ' ';
            else
                cout << '0';
        }
        cout << endl;
    }
}
int easterEgg(){
    ifstream file ("./input");
    string line;
    vector<vector<deque<int>>> map (height, vector<deque<int>> (width, deque<int> ()));
    vector<pair<int,int>> robots;
    int index = 0;
    while (getline(file, line)){
        if (line.empty())
            continue;

        size_t s_pos = line.find(' ');

        string p_string = line.substr(0, s_pos);
        string v_string = line.substr(s_pos+1, line.length()- s_pos -1);
        
        size_t c_pos_p = p_string.find(',');
        size_t c_pos_v = v_string.find(',');

        int x = stoi(p_string.substr(2, c_pos_p -2));
        int y =stoi(p_string.substr(c_pos_p+1, p_string.length() - c_pos_p -1));
        map[y][x].push_back(index++);

        int vx = stoi(v_string.substr(2, c_pos_v -2));
        int vy = stoi(v_string.substr(c_pos_v+1, v_string.length() - c_pos_v -1));
        robots.push_back({vx, vy});
    }
    
    file.close();
    // cout << "t = 0:" << endl;
    // render(map);

    for (int t = 1; t <= 100; t++){
        vector<vector<deque<int>>> newmap (height, vector<deque<int>> (width, deque<int> ()));
        
        for (int i = 0; i < height; i++){
            for (int j = 0; j < width; j++){
                deque<int> rq = map[i][j];
                while (!rq.empty()){
                    int x = j;
                    int y = i;
                    x += (robots[rq.back()].first + width);
                    x %= width;
                    y += (robots[rq.back()].second + height);
                    y %= height;

                    newmap[y][x].push_back(rq.back());
                    rq.pop_back();
                }
            }
        }
        map = newmap;
        if (t  >= 95 && t <=100){
            cout << "t = " << t << ":" << endl;
            render(map);
            cout << endl;
        }
    }

    int ans = 0;
    return ans;
}

int main(){
    int easter_egg = easterEgg();
    cout << easter_egg;
}

File: d14/puzzle2.py


import cv2
import numpy as np

HEIGHT = 103
WIDTH = 101
def render(map: dict, t: int):
    #greyscale black and white image
    img = np.zeros((HEIGHT, WIDTH, 3), dtype=np.uint8)
    
    count = 0
    connect_map = {}
    for item in map:
        img[map[item][1], map[item][0]] = 255
        connect_map[(map[item][0], map[item][1])] = 1
    
    for item in connect_map:
        x = item[0]
        y = item[1]
        if (x, y + 1) in connect_map:
            count += 1
        elif (x, y - 1) in connect_map:
            count += 1
        elif (x + 1, y) in connect_map:
            count += 1
        elif (x - 1, y) in connect_map:
            count += 1
    if count > 100:
        # Paint the black canvas red
        img[np.all(img == [0, 0, 0], axis=-1)] = [0, 0, 255]    
        
    cv2.imwrite(f'./imgs/img{t}.png', img)

def safety_factor():
    with open("./input", "r") as file:
        lines = file.readlines()
    positions = {}
    robots = []
    index = 0
    for line in lines:
        line = line.strip()
        if not line:
            continue

        # Parse position and velocity strings
        s_pos = line.find(' ')
        p_string = line[:s_pos]
        v_string = line[s_pos + 1:]

        c_pos_p = p_string.find(',')
        c_pos_v = v_string.find(',')

        # Extract x, y, vx, vy
        x = int(p_string[2:c_pos_p])
        y = int(p_string[c_pos_p + 1:])
        vx = int(v_string[2:c_pos_v])
        vy = int(v_string[c_pos_v + 1:])

        positions[index] = (x, y)
        robots.append((vx, vy))
        index+= 1
    render(positions, 0)

    test = {}
    idx =0
    for i in range(50):
        for j in range(3):
            test[idx] = (i,j)
            idx += 1
    render(test, -2)

    for t in range(1, 10000):
        print(t)
        newmap = {}
        for item in positions:
            x = positions[item][0]
            y = positions[item][1]
            x = (x + robots[item][0] + WIDTH) % WIDTH
            y = (y + robots[item][1] + HEIGHT) % HEIGHT
            newmap[item] = (x, y)
        positions = newmap
        render(positions, t)
    ans = 0
    return ans

if __name__ == "__main__":
    safety_factor_value = safety_factor()
    print(safety_factor_value)